### PR TITLE
[For v0.19] New draft edit tool for multiple object editing and with all modifiers

### PR DESCRIPTION
**Pull request by Moult**
This is the in-progress PR of creating a new, refactored, edit tool that:

 1. Allows editing of one or more objects at once
 2. Allows you to use all modifiers on individual vertices or edges to modify an existing wire

When complete I am hoping for the current draft edit tool to be deprecated.

Discussion thread: https://forum.freecadweb.org/viewtopic.php?f=23&t=34114

 * It allows you to select multiple objects to edit instead of just one.
 * It highlights the object lines and the points in red.
 * It stays in the mode and allows you to run other modifiers.
 * A very hackish hook into the move modifier is added as a proof of concept.

Thank you for creating a pull request to contribute to FreeCAD! To ease integration, please confirm the following:

- [ ] Branch rebased on latest master `git pull --rebase upstream master`
- [ ] Unit tests confirmed to pass by running `./bin/FreeCAD --run-test 0`
- [ ] Commit message is [well-written](https://chris.beams.io/posts/git-commit/)
- [ ] Commit message includes `issue #<id>` or `fixes #<id>` where `<id>` is the [associated MantisBT](https://freecadweb.org/wiki/tracker#GitHub_and_MantisBT) issue id if one exists

And please remember to update the Wiki with the features added or changed once this PR is merged.

---

### CODE REVIEW

**Comment by Moult** [comment]
PR updated for "move" tool support.

Please watch the video demo below, it shows:

 1. Selecting multiple objects and turning them into an "Edit mode" (the idea is for this mode to highlight more editing handles, allow selection through objects, and show the draft bases of other components in case they are invisible), it also makes it easier to select handles by making them larger and red.
 2. Shows moving a single vertex from one of the selected objects. This is similar to the existing draft edit mode, but more intuitive as it uses the move command so it allows you to pick any arbitrary start and end point.
 3. Shows moving both vertices and edges at the same time from multiple shapes, using the move command.
 4. Shows leaving the "draft edit" mode.

This will make the "stretch" tool and the node move tool for wires and lines obsolete (though not for beziers, arcs, and other special shapes).

https://peertube.social/videos/watch/d39d4f3c-b3c8-4a18-b8c4-6719d0f70f48

**Comment by Moult** [comment]
The last commit basically rounds out this PR in terms of implementing features. It is now possible to:

 - For move, rotate, or scale, choose to modify one or more subelements (vertices / edges) instead of an entire object
 - Highlight a draft object so that it is easier to see / select, even if it is invisible or a base of another component
 - Tab-preselect draft objects to edit them easier with preselection highlighting
 - Fixes the regular scale function which seemed to be just broken (didn't bother splitting out into another PR as I'm refactoring that area anyway, so as to prevent a complex merge conflict)

There is more polish which can be done (in particular, the scale UI code, which is special), as well as more features to be added (array subelements? offset subelements? etc), but I think this is a good stopping point for this PR as it covers move/rotate/scale. Ideally I'd like to wait until v0.18 gets released, merge this PR, do some heavy-testing as well as with carlopav's code, continue cleaning, and then do more PRs to polish/extend this "subelement editing" concept.

What do you reckon?

**Comment by Moult** [comment]
Sorry just a heads up I notice that the "modules" label was added, but I don't think this is correct, due to 0828a4ac8c6152d81a6b7c174facc96781ddea87 and 58df6347b4eba2bef37b96d6c1a4db760658bf85 - it's pretty benign code I think, but I think it does touch core.

File: src/Mod/Draft/DraftTools.py [hunk 1]
```diff
@@ -3967,39 +3967,81 @@ def action(self,arg):
...
+        return command
+
+    def scale_object(self):
+        if self.task.relative.isChecked():
+            self.delta = FreeCAD.DraftWorkingPlane.getGlobalCoords(self.delta)
         objects = '[' + ','.join(['FreeCAD.ActiveDocument.' + object.Name for object in self.selected_objects]) + ']'
         FreeCADGui.addModule("Draft")
-        self.commit(translate("draft","Copy"),
-                    ['Draft.scale('+objects+',delta='+DraftVecUtils.toString(delta)+',center='+DraftVecUtils.toString(self.node[0])+',copy='+str(copy)+',legacy='+str(legacy)+')',
+        self.commit(translate("draft","Copy" if self.task.isCopy.isChecked() else "Scale"),
```
Comment (luzpaz) [inline src/Mod/Draft/DraftTools.py]: @Moult is there a missing `"` specifically before the `if` ?

**Comment by yorikvanhavre** [comment]
Keeping this on hold because of the first line in the PR description above, be sure to notify if there is anything required here..

**Comment by Moult** [comment]
Feel free to merge :) don't think I missed a quote, but I'm kinda on holiday so I don't have a computer to look in detail :)

**Comment by yorikvanhavre** [comment]
Ok will give it a test spin & merge

**Comment by Moult** [comment]
Bump :)

**Comment by yorikvanhavre** [comment]
Okay just gave it a test run now. Nice!! It needs some more work as it fails in a number of cases without warning, for ex. if you select an obj that is not a wire or line. But it doesn't modify the existing behaviour if you don't use the new option, so I think there is no problem in having it merged already...

**Comment by yorikvanhavre** [comment]
Merged with 64d7ed369

**Comment by Moult** [comment]
On Wed, May 08, 2019 at 07:54:54AM -0700, Yorik van Havre wrote:
>    Okay just gave it a test run now. Nice!! It needs some more work as it
>    fails in a number of cases without warning, for ex. if you select an
>    obj that is not a wire or line.

Cheers :) Perhaps you can file bugs and issue them to me and I can fix them as
you discover them?

-- 
Dion Moult

**Comment by yorikvanhavre** [comment]
Sounds like a plan :)

**Pull request closed without merging**